Error message help suggests type ascription on pattern binding

Here's the minimal reproduction:

```rust
fn main() {
    let x = vec![1i32];
    match &x[..] {
        [&v] => {},
        _ => {},
    }
}
```

This produces the following error:

```
error[E0308]: mismatched types
 --> src/main.rs:4:10
  |
4 |         [&v] => {},
  |          ^^ expected i32, found reference
  |
  = note: expected type `i32`
             found type `&_`
  = help: did you mean `v: &i32`?
```

The error rightly points out that we are matching a reference when we actually have the full value itself.

The problem is the help message at the bottom:

```
help: did you mean `v: &i32`?
```

You can't annotate the type of a variable bound in a pattern, so trying this ends up with a syntax error:

```
error: expected one of `,`, `..`, or `@`, found `:`
 --> src/main.rs:4:12
  |
4 |         [&v: &i32] => {},
  |            ^ expected one of `,`, `..`, or `@` here

error: aborting due to previous error
```

That help message should probably not show up in patterns at all.

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -1,8 +1,8 @@
 use crate::check::{FnCtxt, Expectation, Diverges, Needs};
 use crate::check::coercion::CoerceMany;
 use crate::util::nodemap::FxHashMap;
-use errors::Applicability;
-use rustc::hir::{self, PatKind};
+use errors::{Applicability, DiagnosticBuilder};
+use rustc::hir::{self, PatKind, Pat};
 use rustc::hir::def::{Def, CtorKind};
 use rustc::hir::pat_util::EnumerateAndAdjustIterator;
 use rustc::infer;
@@ -377,64 +377,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             // Look for a case like `fn foo(&foo: u32)` and suggest
                             // `fn foo(foo: &u32)`
                             if let Some(mut err) = err {
-                                if let PatKind::Binding(..) = inner.node {
-                                    let parent_id = tcx.hir().get_parent_node_by_hir_id(pat.hir_id);
-                                    let parent = tcx.hir().get_by_hir_id(parent_id);
-                                    debug!("inner {:?} pat {:?} parent {:?}", inner, pat, parent);
-                                    match parent {
-                                        hir::Node::Item(hir::Item {
-                                            node: hir::ItemKind::Fn(..), ..
-                                        }) |
-                                        hir::Node::ForeignItem(hir::ForeignItem {
-                                            node: hir::ForeignItemKind::Fn(..), ..
-                                        }) |
-                                        hir::Node::TraitItem(hir::TraitItem {
-                                            node: hir::TraitItemKind::Method(..), ..
-                                        }) |
-                                        hir::Node::ImplItem(hir::ImplItem {
-                                            node: hir::ImplItemKind::Method(..), ..
-                                        }) => { // this pat is likely an argument
-                                            if let Ok(snippet) = tcx.sess.source_map()
-                                                .span_to_snippet(inner.span)
-                                            { // FIXME: turn into structured suggestion, will need
-                                              // a span that also includes the the arg's type.
-                                                err.help(&format!(
-                                                    "did you mean `{}: &{}`?",
-                                                    snippet,
-                                                    expected,
-                                                ));
-                                            }
-                                        }
-                                        hir::Node::Expr(hir::Expr {
-                                            node: hir::ExprKind::Match(..), ..
-                                        }) => { // rely on match ergonomics
-                                            if let Ok(snippet) = tcx.sess.source_map()
-                                                .span_to_snippet(inner.span)
-                                            {
-                                                err.span_suggestion(
-                                                    pat.span,
-                                                    "you can rely on match ergonomics and remove \
-                                                     the explicit borrow",
-                                                    snippet,
-                                                    Applicability::MaybeIncorrect,
-                                                );
-                                            }
-                                        }
-                                        hir::Node::Pat(_) => {  // nested `&&pat`
-                                            if let Ok(snippet) = tcx.sess.source_map()
-                                                .span_to_snippet(inner.span)
-                                            {
-                                                err.span_suggestion(
-                                                    pat.span,
-                                                    "you can probaly remove the explicit borrow",
-                                                    snippet,
-                                                    Applicability::MaybeIncorrect,
-                                                );
-                                            }
-                                        }
-                                        _ => {} // don't provide suggestions in other cases #55175
-                                    }
-                                }
+                                self.borrow_pat_suggestion(&mut err, &pat, &inner, &expected);
                                 err.emit();
                             }
                             (rptr_ty, inner_ty)
@@ -564,6 +507,59 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         // bound in a closure signature, and that detection gets thrown
         // off when we substitute fresh region variables here to enable
         // subtyping.
+    }
+
+    fn borrow_pat_suggestion(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        pat: &Pat,
+        inner: &Pat,
+        expected: &Ty<'tcx>,
+    ) {
+        let tcx = self.tcx;
+        if let PatKind::Binding(..) = inner.node {
+            let parent_id = tcx.hir().get_parent_node_by_hir_id(pat.hir_id);
+            let parent = tcx.hir().get_by_hir_id(parent_id);
+            debug!("inner {:?} pat {:?} parent {:?}", inner, pat, parent);
+            match parent {
+                hir::Node::Item(hir::Item { node: hir::ItemKind::Fn(..), .. }) |
+                hir::Node::ForeignItem(hir::ForeignItem {
+                    node: hir::ForeignItemKind::Fn(..), ..
+                }) |
+                hir::Node::TraitItem(hir::TraitItem { node: hir::TraitItemKind::Method(..), .. }) |
+                hir::Node::ImplItem(hir::ImplItem { node: hir::ImplItemKind::Method(..), .. }) => {
+                    // this pat is likely an argument
+                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
+                        // FIXME: turn into structured suggestion, will need a span that also
+                        // includes the the arg's type.
+                        err.help(&format!("did you mean `{}: &{}`?", snippet, expected));
+                    }
+                }
+                hir::Node::Expr(hir::Expr { node: hir::ExprKind::Match(..), .. }) => {
+                    // rely on match ergonomics
+                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
+                        err.span_suggestion(
+                            pat.span,
+                            "you can rely on match ergonomics and remove the explicit borrow",
+                            snippet,
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                }
+                hir::Node::Pat(_) => {
+                    // nested `&&pat`
+                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
+                        err.span_suggestion(
+                            pat.span,
+                            "you can probably remove the explicit borrow",
+                            snippet,
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                }
+                _ => {} // don't provide suggestions in other cases #55175
+            }
+        }
     }
 
     pub fn check_dereferencable(&self, span: Span, expected: Ty<'tcx>, inner: &hir::Pat) -> bool {

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -535,19 +535,9 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         err.help(&format!("did you mean `{}: &{}`?", snippet, expected));
                     }
                 }
-                hir::Node::Expr(hir::Expr { node: hir::ExprKind::Match(..), .. }) => {
-                    // rely on match ergonomics
-                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
-                        err.span_suggestion(
-                            pat.span,
-                            "you can rely on match ergonomics and remove the explicit borrow",
-                            snippet,
-                            Applicability::MaybeIncorrect,
-                        );
-                    }
-                }
+                hir::Node::Expr(hir::Expr { node: hir::ExprKind::Match(..), .. }) |
                 hir::Node::Pat(_) => {
-                    // nested `&&pat`
+                    // rely on match ergonomics or it might be nested `&&pat`
                     if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(inner.span) {
                         err.span_suggestion(
                             pat.span,

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -378,12 +378,25 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             // `fn foo(foo: &u32)`
                             if let Some(mut err) = err {
                                 if let PatKind::Binding(..) = inner.node {
-                                    if let Ok(snippet) = tcx.sess.source_map()
-                                                                    .span_to_snippet(pat.span)
-                                    {
-                                        err.help(&format!("did you mean `{}: &{}`?",
-                                                            &snippet[1..],
-                                                            expected));
+                                    let parent_id = tcx.hir().get_parent_node_by_hir_id(pat.hir_id);
+                                    let parent = tcx.hir().get_by_hir_id(parent_id);
+                                    match parent {
+                                        hir::Node::Item(_) |
+                                        hir::Node::ForeignItem(_) |
+                                        hir::Node::TraitItem(_) |
+                                        hir::Node::ImplItem(_)  => { // this pat is an argument
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(pat.span)
+                                            { // FIXME: turn into structured suggestion, will need
+                                              // a span that also includes the the type.
+                                                err.help(&format!(
+                                                    "did you mean `{}: &{}`?",
+                                                    &snippet[1..],
+                                                    expected,
+                                                ));
+                                            }
+                                        }
+                                        _ => {} // don't provide the suggestion from above #55175
                                     }
                                 }
                                 err.emit();

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -380,23 +380,59 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                                 if let PatKind::Binding(..) = inner.node {
                                     let parent_id = tcx.hir().get_parent_node_by_hir_id(pat.hir_id);
                                     let parent = tcx.hir().get_by_hir_id(parent_id);
+                                    debug!("inner {:?} pat {:?} parent {:?}", inner, pat, parent);
                                     match parent {
-                                        hir::Node::Item(_) |
-                                        hir::Node::ForeignItem(_) |
-                                        hir::Node::TraitItem(_) |
-                                        hir::Node::ImplItem(_)  => { // this pat is an argument
+                                        hir::Node::Item(hir::Item {
+                                            node: hir::ItemKind::Fn(..), ..
+                                        }) |
+                                        hir::Node::ForeignItem(hir::ForeignItem {
+                                            node: hir::ForeignItemKind::Fn(..), ..
+                                        }) |
+                                        hir::Node::TraitItem(hir::TraitItem {
+                                            node: hir::TraitItemKind::Method(..), ..
+                                        }) |
+                                        hir::Node::ImplItem(hir::ImplItem {
+                                            node: hir::ImplItemKind::Method(..), ..
+                                        }) => { // this pat is likely an argument
                                             if let Ok(snippet) = tcx.sess.source_map()
-                                                .span_to_snippet(pat.span)
+                                                .span_to_snippet(inner.span)
                                             { // FIXME: turn into structured suggestion, will need
-                                              // a span that also includes the the type.
+                                              // a span that also includes the the arg's type.
                                                 err.help(&format!(
                                                     "did you mean `{}: &{}`?",
-                                                    &snippet[1..],
+                                                    snippet,
                                                     expected,
                                                 ));
                                             }
                                         }
-                                        _ => {} // don't provide the suggestion from above #55175
+                                        hir::Node::Expr(hir::Expr {
+                                            node: hir::ExprKind::Match(..), ..
+                                        }) => { // rely on match ergonomics
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            {
+                                                err.span_suggestion(
+                                                    pat.span,
+                                                    "you can rely on match ergonomics and remove \
+                                                     the explicit borrow",
+                                                    snippet,
+                                                    Applicability::MaybeIncorrect,
+                                                );
+                                            }
+                                        }
+                                        hir::Node::Pat(_) => {  // nested `&&pat`
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            {
+                                                err.span_suggestion(
+                                                    pat.span,
+                                                    "you can probaly remove the explicit borrow",
+                                                    snippet,
+                                                    Applicability::MaybeIncorrect,
+                                                );
+                                            }
+                                        }
+                                        _ => {} // don't provide suggestions in other cases #55175
                                     }
                                 }
                                 err.emit();

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -514,7 +514,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         err: &mut DiagnosticBuilder<'_>,
         pat: &Pat,
         inner: &Pat,
-        expected: &Ty<'tcx>,
+        expected: Ty<'tcx>,
     ) {
         let tcx = self.tcx;
         if let PatKind::Binding(..) = inner.node {

--- a/src/test/ui/destructure-trait-ref.stderr
+++ b/src/test/ui/destructure-trait-ref.stderr
@@ -24,7 +24,6 @@ LL |     let &&x = &1isize as &T;
    |
    = note: expected type `dyn T`
               found type `&_`
-   = help: did you mean `x: &dyn T`?
 
 error[E0308]: mismatched types
   --> $DIR/destructure-trait-ref.rs:36:11
@@ -34,7 +33,6 @@ LL |     let &&&x = &(&1isize as &T);
    |
    = note: expected type `dyn T`
               found type `&_`
-   = help: did you mean `x: &dyn T`?
 
 error[E0308]: mismatched types
   --> $DIR/destructure-trait-ref.rs:41:13

--- a/src/test/ui/destructure-trait-ref.stderr
+++ b/src/test/ui/destructure-trait-ref.stderr
@@ -23,7 +23,7 @@ LL |     let &&x = &1isize as &T;
    |          ^^
    |          |
    |          expected trait T, found reference
-   |          help: you can probaly remove the explicit borrow: `x`
+   |          help: you can probably remove the explicit borrow: `x`
    |
    = note: expected type `dyn T`
               found type `&_`
@@ -35,7 +35,7 @@ LL |     let &&&x = &(&1isize as &T);
    |           ^^
    |           |
    |           expected trait T, found reference
-   |           help: you can probaly remove the explicit borrow: `x`
+   |           help: you can probably remove the explicit borrow: `x`
    |
    = note: expected type `dyn T`
               found type `&_`

--- a/src/test/ui/destructure-trait-ref.stderr
+++ b/src/test/ui/destructure-trait-ref.stderr
@@ -20,7 +20,10 @@ error[E0308]: mismatched types
   --> $DIR/destructure-trait-ref.rs:31:10
    |
 LL |     let &&x = &1isize as &T;
-   |          ^^ expected trait T, found reference
+   |          ^^
+   |          |
+   |          expected trait T, found reference
+   |          help: you can probaly remove the explicit borrow: `x`
    |
    = note: expected type `dyn T`
               found type `&_`
@@ -29,7 +32,10 @@ error[E0308]: mismatched types
   --> $DIR/destructure-trait-ref.rs:36:11
    |
 LL |     let &&&x = &(&1isize as &T);
-   |           ^^ expected trait T, found reference
+   |           ^^
+   |           |
+   |           expected trait T, found reference
+   |           help: you can probaly remove the explicit borrow: `x`
    |
    = note: expected type `dyn T`
               found type `&_`

--- a/src/test/ui/mismatched_types/issue-38371.stderr
+++ b/src/test/ui/mismatched_types/issue-38371.stderr
@@ -16,7 +16,6 @@ LL | fn agh(&&bar: &u32) {
    |
    = note: expected type `u32`
               found type `&_`
-   = help: did you mean `bar: &u32`?
 
 error[E0308]: mismatched types
   --> $DIR/issue-38371.rs:21:8

--- a/src/test/ui/mismatched_types/issue-38371.stderr
+++ b/src/test/ui/mismatched_types/issue-38371.stderr
@@ -12,7 +12,10 @@ error[E0308]: mismatched types
   --> $DIR/issue-38371.rs:18:9
    |
 LL | fn agh(&&bar: &u32) {
-   |         ^^^^ expected u32, found reference
+   |         ^^^^
+   |         |
+   |         expected u32, found reference
+   |         help: you can probaly remove the explicit borrow: `bar`
    |
    = note: expected type `u32`
               found type `&_`

--- a/src/test/ui/mismatched_types/issue-38371.stderr
+++ b/src/test/ui/mismatched_types/issue-38371.stderr
@@ -15,7 +15,7 @@ LL | fn agh(&&bar: &u32) {
    |         ^^^^
    |         |
    |         expected u32, found reference
-   |         help: you can probaly remove the explicit borrow: `bar`
+   |         help: you can probably remove the explicit borrow: `bar`
    |
    = note: expected type `u32`
               found type `&_`

--- a/src/test/ui/suggestions/match-ergonomics.rs
+++ b/src/test/ui/suggestions/match-ergonomics.rs
@@ -37,4 +37,5 @@ fn main() {
         v => {},
         _ => {},
     }
+    if let [&v] = &x[..] {} //~ ERROR mismatched types
 }

--- a/src/test/ui/suggestions/match-ergonomics.rs
+++ b/src/test/ui/suggestions/match-ergonomics.rs
@@ -1,0 +1,40 @@
+fn main() {
+    let x = vec![1i32];
+    match &x[..] {
+        [&v] => {}, //~ ERROR mismatched types
+        _ => {},
+    }
+    match x {
+        [&v] => {}, //~ ERROR expected an array or slice
+        _ => {},
+    }
+    match &x[..] {
+        [v] => {},
+        _ => {},
+    }
+    match &x[..] {
+        &[v] => {},
+        _ => {},
+    }
+    match x {
+        [v] => {}, //~ ERROR expected an array or slice
+        _ => {},
+    }
+    let y = 1i32;
+    match &y {
+        &v => {},
+        _ => {},
+    }
+    match y {
+        &v => {}, //~ ERROR mismatched types
+        _ => {},
+    }
+    match &y {
+        v => {},
+        _ => {},
+    }
+    match y {
+        v => {},
+        _ => {},
+    }
+}

--- a/src/test/ui/suggestions/match-ergonomics.stderr
+++ b/src/test/ui/suggestions/match-ergonomics.stderr
@@ -26,14 +26,13 @@ error[E0308]: mismatched types
   --> $DIR/match-ergonomics.rs:29:9
    |
 LL |         &v => {},
-   |         ^^ expected i32, found reference
+   |         ^^
+   |         |
+   |         expected i32, found reference
+   |         help: you can probably remove the explicit borrow: `v`
    |
    = note: expected type `i32`
               found type `&_`
-help: you can rely on match ergonomics and remove the explicit borrow
-   |
-LL |         v => {},
-   |         ^
 
 error[E0308]: mismatched types
   --> $DIR/match-ergonomics.rs:40:13

--- a/src/test/ui/suggestions/match-ergonomics.stderr
+++ b/src/test/ui/suggestions/match-ergonomics.stderr
@@ -35,7 +35,19 @@ help: you can rely on match ergonomics and remove the explicit borrow
 LL |         v => {},
    |         ^
 
-error: aborting due to 4 previous errors
+error[E0308]: mismatched types
+  --> $DIR/match-ergonomics.rs:40:13
+   |
+LL |     if let [&v] = &x[..] {}
+   |             ^^
+   |             |
+   |             expected i32, found reference
+   |             help: you can probably remove the explicit borrow: `v`
+   |
+   = note: expected type `i32`
+              found type `&_`
+
+error: aborting due to 5 previous errors
 
 Some errors have detailed explanations: E0308, E0529.
 For more information about an error, try `rustc --explain E0308`.

--- a/src/test/ui/suggestions/match-ergonomics.stderr
+++ b/src/test/ui/suggestions/match-ergonomics.stderr
@@ -5,7 +5,7 @@ LL |         [&v] => {},
    |          ^^
    |          |
    |          expected i32, found reference
-   |          help: you can probaly remove the explicit borrow: `v`
+   |          help: you can probably remove the explicit borrow: `v`
    |
    = note: expected type `i32`
               found type `&_`

--- a/src/test/ui/suggestions/match-ergonomics.stderr
+++ b/src/test/ui/suggestions/match-ergonomics.stderr
@@ -1,0 +1,41 @@
+error[E0308]: mismatched types
+  --> $DIR/match-ergonomics.rs:4:10
+   |
+LL |         [&v] => {},
+   |          ^^
+   |          |
+   |          expected i32, found reference
+   |          help: you can probaly remove the explicit borrow: `v`
+   |
+   = note: expected type `i32`
+              found type `&_`
+
+error[E0529]: expected an array or slice, found `std::vec::Vec<i32>`
+  --> $DIR/match-ergonomics.rs:8:9
+   |
+LL |         [&v] => {},
+   |         ^^^^ pattern cannot match with input type `std::vec::Vec<i32>`
+
+error[E0529]: expected an array or slice, found `std::vec::Vec<i32>`
+  --> $DIR/match-ergonomics.rs:20:9
+   |
+LL |         [v] => {},
+   |         ^^^ pattern cannot match with input type `std::vec::Vec<i32>`
+
+error[E0308]: mismatched types
+  --> $DIR/match-ergonomics.rs:29:9
+   |
+LL |         &v => {},
+   |         ^^ expected i32, found reference
+   |
+   = note: expected type `i32`
+              found type `&_`
+help: you can rely on match ergonomics and remove the explicit borrow
+   |
+LL |         v => {},
+   |         ^
+
+error: aborting due to 4 previous errors
+
+Some errors have detailed explanations: E0308, E0529.
+For more information about an error, try `rustc --explain E0308`.

